eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs

```diff
@@ -501,8 +504,48 @@ expectPhase2Invalid ::
   ) =>
   Tx era ->
   ImpTestM era ()
-expectPhase2Invalid tx = do
+expectPhase2Invalid = void . submitPhase2InvalidTx
+
+submitPhase2InvalidTx ::
+  ( HasCallStack
+  , AlonzoEraImp era
+  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
+  ) =>
+  Tx era ->
+  ImpTestM era (Tx era)
+submitPhase2InvalidTx tx = do
   (predFailure, fixedUpTx) <- expectLeft =<< trySubmitTx tx
   scriptPredicateFailure <- impScriptPredicateFailure fixedUpTx
   predFailure `shouldBeExpr` pure (injectFailure scriptPredicateFailure)
-  withNoFixup $ submitTx_ $ fixedUpTx & isValidTxL .~ IsValid False
+  withNoFixup $ submitTx $ fixedUpTx & isValidTxL .~ IsValid False
+
+expectTxSuccess ::
+  ( HasCallStack
+  , AlonzoEraImp era
+  ) =>
+  Tx era -> ImpTestM era ()
+expectTxSuccess tx = do
+  utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
+  let inputs = Set.toList $ tx ^. bodyTxL . inputsTxBodyL
+      outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
+  impAnn "Inputs should be gone from UTxO" $
+    expectUTxOContent utxo [(txIn, isNothing) | txIn <- inputs]
+  impAnn "Outputs should be in UTxO" $
+    expectUTxOContent utxo [(txIn, (== Just txOut)) | (txIn, txOut) <- outputs]
+
+expectTxPhase2Failure ::
+  ( HasCallStack
+  , AlonzoEraImp era
+  ) =>
+  Tx era -> ImpTestM era ()
+expectTxPhase2Failure tx = do
+  utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
+  let inputs = tx ^. bodyTxL . inputsTxBodyL
+      collaterals = tx ^. bodyTxL . collateralInputsTxBodyL
+      outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
+  impAnn "Non-collateral inputs should still be in UTxO" $
+    expectUTxOContent utxo [(txIn, isJust) | txIn <- Set.toList $ inputs \\ collaterals]
+  impAnn "Collateral inputs should not be in UTxO" $
+    expectUTxOContent utxo [(txIn, isNothing) | txIn <- Set.toList collaterals]
+  impAnn "Outputs should not be in UTxO" $
+    expectUTxOContent utxo [(txIn, isNothing) | (txIn, _txOut) <- outputs]
```

Comment:
This is morally one function, because in both cases it is submitted to the chain successfully.
```suggestion
expectTxSuccess ::
  ( HasCallStack
  , AlonzoEraImp era
  ) =>
  Tx era -> ImpTestM era ()
expectTxSuccess tx 
  | tx ^. isValidTxL == IsValid True = do
    utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
    let inputs = Set.toList $ tx ^. bodyTxL . inputsTxBodyL
        outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
    impAnn "Inputs should be gone from UTxO" $
      expectUTxOContent utxo [(txIn, isNothing) | txIn <- inputs]
    impAnn "Outputs should be in UTxO" $
      expectUTxOContent utxo [(txIn, (== Just txOut)) | (txIn, txOut) <- outputs]
  | otherwise = do
    utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
    let inputs = tx ^. bodyTxL . inputsTxBodyL
        collaterals = tx ^. bodyTxL . collateralInputsTxBodyL
        outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
    impAnn "Non-collateral inputs should still be in UTxO" $
      expectUTxOContent utxo [(txIn, isJust) | txIn <- Set.toList $ inputs \\ collaterals]
    impAnn "Collateral inputs should not be in UTxO" $
      expectUTxOContent utxo [(txIn, isNothing) | txIn <- Set.toList collaterals]
    impAnn "Outputs should not be in UTxO" $
      expectUTxOContent utxo [(txIn, isNothing) | (txIn, _txOut) <- outputs]
```

In the matter of fact, reviewing this I had an idea. We can make this an `EraImp` type class function and execute it in the successful case of `trySubmitTx`. This would make sure we always check these invarianta for all of our imp tests.

It has to be a type class function because different eras will have different things we can be checking. That being said, you don't have to make this change in this PR, we can do it later in a separate PR

Reply:
I love this suggestion. Done.